File: cs/type-safe-builder/StaticOptional.cs
namespace type_safe_builder
{
	public interface StaticOptional<Type> {}

	public struct StaticSome<Type> : StaticOptional<Type>
	{
		public Type Element { get;  }

		public StaticSome(Type element)
		{
			Element = element;
		}
	}

	public struct StaticNone<Type> : StaticOptional<Type> {}
}

File: rust/src/static_optional.rs
use std::marker::PhantomData;

pub trait StaticOptional {
	type Element;

	fn some(element: Self::Element) -> StaticSome<Self::Element> {
		StaticSome(element)
	}

	fn none() -> StaticNone<Self::Element> {
		StaticNone::default()
	}
}

pub struct StaticSome<Element>(pub Element);

impl<Element> StaticOptional for StaticSome<Element> {
	type Element = Element;
}

impl<Element> StaticOptional for StaticNone<Element> {
	type Element = Element;
}

pub struct StaticNone<Element> {
	_element: PhantomData<Element>,
}

impl<Element> Default for StaticNone<Element> {
	fn default() -> Self {
		Self {
			_element: Default::default(),
		}
	}
}
